### PR TITLE
chore: Lock main dependencies for CI

### PR DESCRIPTION
**Pull request by AdrianoKF**
This PR adds a lock file for the main dependencies, `requirements.txt`, obtained through pip-compile, similar to the existing dev dependencies.

The lockfile for dev dependencies has been renamed to `requirements-dev.txt` for consistency.

The contribution guide and CI actions have been updated to reflect the new changes.

Also, this change adds pre-commit hooks for automatically maintaining the lockfiles via `pip-compile`.

⚠️ Due to a breaking API change in `lakefs-client==0.108.0`, the dependency has been pinned at `0.107.1` for now.

### CODE REVIEW

File: .github/workflows/python.yaml [hunk 1]
```diff
@@ -38,11 +38,13 @@ jobs:
...
         cache-dependency-path: |
-          dev-deps.lock
+          requirements.txt
+          requirements-dev.txt
           pyproject.toml
     - name: Install dependencies
       run: |
-        pip install -r dev-deps.lock -e .
+        pip install -r requirements.txt -r requirements-dev.txt
+        pip install -e . --no-deps
```
Comment (nicholasjng) [inline .github/workflows/python.yaml:47]: Nice! Much more predictable as two commands imo

**Review by nicholasjng — COMMENTED** [review]
Looks great! Let's see about the new client version, also thank you for adding discussions.

**Comment by AdrianoKF** [comment]
> Looks great! Let's see about the new client version, also thank you for adding discussions.

We could also pin lakeFS to 0.107.0 for the time being and deal with the API change in a follow-up PR for clarity.

**Review by nicholasjng — APPROVED** [review]

**Comment by AdrianoKF** [comment]
Follow-up issue for the breaking API change: #47

**Pull request merged**